### PR TITLE
feat: analytics tracking sdk integration

### PR DESCRIPTION
**Pull request by bozzelliandrea**
### Ticket:

- [21905](https://dev.azure.com/OGCIO-Digital-Services/Digital%20Services%20Programme/_workitems/edit/21905)

### Description

Added on payments BB (frontend and api) the integration with analytics sdk.

quick note: the package `/analytics-sdk/` is temporary and we will replace it with the npm dependency as soon as it is ready
 
## Type

- [ ] **Dependency upgrade**
- [ ] **Bug fix**
- [X] **New feature**
- [ ] **Dev change**
- [ ] **Additional tests**
- [ ] **Documentation**
- [ ] **Other**

### Checklist:

- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] This change might impact the developer experience of others and should be communicated

### Screenshots:

N/A

### CODE REVIEW

File: packages/analytics/components/AnalyticsEventSender.tsx [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+  useEffect(() => {
+    if (isInitialLoad.current) {
+      isInitialLoad.current = false;
+      return;
+    }
+    if(window["_paq"]) {
+      window["_paq"].push(["setCustomUrl", window.location.origin + pathname]);
+      window["_paq"].push(["trackPageView"]);
+    }
```
Comment (andreacappadona17) [inline packages/analytics/components/AnalyticsEventSender.tsx:19]: Once ready, we want to use the client side methods from the SDK.
Please add a comment with a TODO to ensure we do not lose track of this

Comment (bozzelliandrea) [reply packages/analytics/components/AnalyticsEventSender.tsx:20]: comment added

File: apps/payments-api/routes/auditLogs/index.ts [hunk 2]
```diff
@@ -56,13 +56,21 @@ export default async function auditLogs(app: FastifyInstance) {
...
+        app.buildingBlocks.sdk.analytics
+          .trackEvent({
+            event: {
+              category: "API",
+              name: "AuditLogs",
+              action: "POST",
+              value: 1,
+            },
+          })
+          .then((_) => console.log("Track Success "));
```
Comment (andreacappadona17) [inline apps/payments-api/routes/auditLogs/index.ts:69]:
```suggestion
          .then((_) => console.log("Track Success"));
```

File: apps/payments/app/components/AnalyticsTracker.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import { getBuildingBlockSDK } from "@ogcio/building-blocks-sdk";
+
+const buildingBlocksFrontend = getBuildingBlockSDK({
+  services: {
+    analytics: {
+      baseUrl: process.env.NEXT_PUBLIC_ANALYTICS_URL!,
+      matomoToken: process.env.NEXT_PUBLIC_ANALYTICS_MATOMO_TOKEN,
+      trackingWebsiteId: process.env.NEXT_PUBLIC_ANALYTICS_WEBSITE_ID,
+      organizationId: process.env.NEXT_PUBLIC_AUTH_ORGANIZATION_ID!,
+    },
```
Comment (andreacappadona17) [inline apps/payments/app/components/AnalyticsTracker.tsx:14]: Avoid exposing Matomo Token on client side

```suggestion
    analytics: {
      baseUrl: process.env.NEXT_PUBLIC_ANALYTICS_URL!,
      trackingWebsiteId: process.env.NEXT_PUBLIC_ANALYTICS_WEBSITE_ID,
      organizationId: process.env.NEXT_PUBLIC_AUTH_ORGANIZATION_ID!,
    },
```

File: apps/payments/app/components/AnalyticsTracker.tsx [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+export default function AnalyticsTracker({ userId }: AnalyticsTrackerProps) {
+  const pathname = usePathname();
+  const isInitialLoad = useRef(true);
+  const isInitialized = useRef(false);
+
+  async function initAnalytics(userId?: string) {
+    await buildingBlocksFrontend.analytics.initClientTracker({
+      user_id: userId,
```
Comment (andreacappadona17) [inline apps/payments/app/components/AnalyticsTracker.tsx:35]: Ensure GDPR compliancy if we keep tracking the userId.

File: packages/analytics/components/AnalyticsTracker.tsx [hunk 5]
```diff
@@ -20,18 +20,12 @@ interface AnalyticsTrackerProps {
...
-  if (!ANALYTICS_WEBSITE_ID) {
-    return null;
-  }
-
-  const ogcioAnalyticsSDK = new OGCIO();
-  await ogcioAnalyticsSDK.auth(authConfig);
-  const trackingScript = await ogcioAnalyticsSDK.analytics.getTrackingCode(process.env.ANALYTICS_WEBSITE_ID, {});
-  return (<>
-    <Script>{trackingScript.data?.data.script}</Script>
-  </>);
```
Comment (andreacappadona17) [inline packages/analytics/components/AnalyticsTracker.tsx]: Other projects in the repo may be still using this.
Ensure that there is no broken dependencies

File: apps/payments/app/components/AnalyticsTracker.tsx
Comment (andreacappadona17) [inline apps/payments/app/components/AnalyticsTracker.tsx:1]: To be moved outside of Payments once we integrate with the other BBs of the repo.

**Comment by andreacappadona17** [comment]
@bozzelliandrea added a couple comments to the review.
Please also make sure that all the conflicts with dev branch are solved.

File: packages/analytics/components/AnalyticsTracker.tsx (hunk 5 above)
Comment (bozzelliandrea) [reply packages/analytics/components/AnalyticsTracker.tsx]: @andreacappadona17 added a refactor for the [AnalyticsTracker.tsx](https://github.com/ogcio/life-events/pull/1028/files#diff-5b65df2e59061ddc101ff9b54f01ce117c22490daf34fb89194037fe03bbf602) component, supporting an hybrid solution with old env variables.
We can use this until we finish the update for every submodule

File: apps/payments-api/routes/auditLogs/index.ts [hunk 6]
```diff
@@ -56,6 +56,22 @@ export default async function auditLogs(app: FastifyInstance) {
...
+              category: "API",
+              name: "AuditLogs",
+              action: "POST",
+              value: 1,
+            },
+          })
+          .then((_) => console.log("Track Success"));
+      } catch (err) {
+        //TODO: next analytics version will support silence errors to avoid app crashes
+        console.error(err);
```
Comment (SamSalvatico) [inline apps/payments-api/routes/auditLogs/index.ts:72]:
```suggestion
        app.log.error(err);
```

File: apps/payments-api/routes/auditLogs/index.ts [hunk 7]
```diff
@@ -56,6 +56,22 @@ export default async function auditLogs(app: FastifyInstance) {
...
+        app.buildingBlocks.sdk.analytics
+          .trackEvent({
+            event: {
+              category: "API",
+              name: "AuditLogs",
+              action: "POST",
+              value: 1,
+            },
+          })
+          .then((_) => console.log("Track Success"));
```
Comment (SamSalvatico) [inline apps/payments-api/routes/auditLogs/index.ts:69]:
```suggestion
          .then((_) => app.log.debug("Track Success"));
```

Comment (bozzelliandrea) [reply apps/payments-api/routes/auditLogs/index.ts]: @SamSalvatico fixed with app log

**Review by SamSalvatico — DISMISSED** [review]
👌

**Review by andreacappadona17 — APPROVED** [review]
LGTM

Sync with @williamwd and @alfonsograziano for the env variables values configuration on infrastructure and logto seeder before merging

File: apps/payments-api/routes/auditLogs/index.ts [hunk 8]
```diff
@@ -56,6 +56,22 @@ export default async function auditLogs(app: FastifyInstance) {
...
+              action: "POST",
+              value: 1,
+            },
+          })
+          .then(() => app.log.debug("Track Success"));
+      } catch (err) {
+        //TODO: next analytics version will support silence errors to avoid app crashes
+        app.log.error(err);
+      }
+
```
Comment (alewin) [inline apps/payments-api/routes/auditLogs/index.ts:74]:
```suggestion
       const { data, error } =  app.buildingBlocks.sdk.analytics
          .track.event({
            event: {
              category: "PAYMENTS_API",
              name: "AuditLogs",
              action: "POST",
              value: 1,
            },
          });
          
          if(!error) app.log.debug("Track Success");
          else  app.log.error(err);

```

File: packages/analytics/__tests__/matomo.test.ts
Comment (alewin) [inline packages/analytics/__tests__/matomo.test.ts:1]: is this file used?

File: apps/payments/.env.sample [hunk 9]
```diff
@@ -37,3 +37,19 @@ ENVIRONMENT=DEV
...
+AUTH_APP_ID=auth_app_id
+AUTH_APP_SECRET=auth_app_secret
+
+ANALYTICS_URL=http://localhost:8075
+ANALYTICS_WEBSITE_ID=website_id
+ANALYTICS_MATOMO_TOKEN=
+
+NEXT_PUBLIC_ANALYTICS_URL=http://localhost:8075
+NEXT_PUBLIC_ANALYTICS_WEBSITE_ID=website_id
+NEXT_PUBLIC_AUTH_ORGANIZATION_ID=org_id
```
Comment (alewin) [inline apps/payments/.env.sample:55]: in the previus `.env` we defined the `NEXT_PUBLIC_AUTH_ORGANIZATION_ID` as `ogcio` why here no?

File: packages/analytics/components/AnalyticsTracker.tsx [hunk 10]
```diff
@@ -1,12 +1,27 @@
...
+  services: {
+    analytics: {
+      baseUrl:
+        process.env.NEXT_PUBLIC_ANALYTICS_URL ??
+        (process.env.NEXT_PUBLIC_MATOMO_PROTOCOL ?? "http") +
+          process.env.NEXT_PUBLIC_MATOMO_URL,
+      trackingWebsiteId:
+        process.env.NEXT_PUBLIC_ANALYTICS_WEBSITE_ID ??
+        process.env.NEXT_PUBLIC_MATOMO_SITE_ID,
+      organizationId: process.env.NEXT_PUBLIC_AUTH_ORGANIZATION_ID ?? "ogcio",
```
Comment (alewin) [inline packages/analytics/components/AnalyticsTracker.tsx:17]:
```suggestion
      organizationId: process.env.NEXT_PUBLIC_AUTH_ORGANIZATION_ID,
```

File: packages/analytics/components/AnalyticsTracker.tsx [hunk 11]
```diff
@@ -15,64 +30,39 @@ interface AnalyticsTrackerProps {
...
-export default function AnalyticsTracker({
-  userId,
-  customDimensions,
-}: AnalyticsTrackerProps) {
-  // avoid tracking if no MATOMO_URL or MATOMO_SITE_ID
-  if (!MATOMO_URL || !MATOMO_SITE_ID) {
-    return null;
-  }
-
+export default function AnalyticsTracker({ userId }: AnalyticsTrackerProps) {
```
Comment (alewin) [inline packages/analytics/components/AnalyticsTracker.tsx:37]: we should also read the `customDimensions`

File: apps/payments-api/package.json [hunk 12]
```diff
@@ -40,6 +40,7 @@
     "@fastify/swagger-ui": "^3.0.0",
     "@fastify/type-provider-typebox": "^4.0.0",
     "@fastify/under-pressure": "^8.3.0",
+    "@ogcio/building-blocks-sdk": "^0.0.21",
```
Comment (alewin) [inline apps/payments-api/package.json:43]: i created a PR with the new version `0.0.23` https://github.com/ogcio/building-blocks-sdk/pull/57 with some improvements and fixes about the `analytics` building block

File: packages/analytics/__tests__/matomo.test.ts
Comment (bozzelliandrea) [reply packages/analytics/__tests__/matomo.test.ts:1]: no, in fact I deleted it

File: apps/web/app/[locale]/(public)/layout.tsx [hunk 13]
```diff
@@ -86,7 +86,7 @@ export default async function RootLayout({
       >
         <AnalyticsTracker
           userId={userId}
-          customDimensions={{ dimension1: verificationLevel }}
+          customDimensions={{ dimension1: verificationLevel + "" }}
```
[the remark this answers is not in the document]
Comment (bozzelliandrea) [reply apps/web/app/[locale]/(public)/layout.tsx:89]: verificationLevel is a number, and dimensionValue can only be a string.

With the new analytics release can also be a number?

**Review by alewin — APPROVED** [review]

**Pull request merged**